Migrate import store from Flux to Redux

talldan opened the pull request:
The imports store was already written in a fairly redux-y way. This PR migrates it fully to flux. It:
- Adds `state/imports/reducer.js`, which is a reworking of `lib/importer/store.js`. The new version has immutable stripped out, using a mixture of modern js and lodash to achieve immutability.
- Adds `state/imports/actions.js`, which replaces `lib/importer/actions.js`. Functions have been modified so that the dispatcher is passed in as an argument.
- Reworks any connected components to use redux actions and react-redux `connect`.

### To Test: ###
**WordPress Importer**
1. Export WordPress site data by going to settings > export > export all
2. Unzip the the export and find the xml file in the unzipped folder. 
3. Preferably using a different WordPress site, go to settings > import
5. Import .xml file using the WordPress importer
6. Confirm that the import process completes successfully
7. Make a copy of the xml file from step 2 and rename the copy to something like `wordpress-export-with-error.xml`
8. Edit the `wordpress-export-with-error.xml` file you just created, and remove the line `<wp:wxr_version>1.2</wp:wxr_version>`. This second export file will be used to test errors in the importing process.
9. Now import the `wordpress-export-with-error.xml` file again using the WordPress importer.
10. Confirm that you see an error message `There was a problem with your WordPress import file. It doesn't seem to be in a valid format.`

**Medium Importer**
1. Export data from a medium site
2. Back in WordPress, go to settings > import
3. Click 'Start import' on the Medium importer
4. Import .zip file downloaded from medium
5. Confirm that the import process completes successfully

alisterscott commented:
Hi @talldan - thanks for your contribution - is this one ready for review?

talldan commented:
Hi @alisterscott - I'm adding some tests today, and then it should be ready. This is part of a trial, btw, so I've been discussing the PR in slack in the delta-samus channel.

spen commented:
Hi @talldan!

Would you mind adding a test plan detailing how best to cover the behaviour that could be effected by these changes? There's no behavioural change but I'm keen to run through the flows that would be at risk of regression.

Thanks!

talldan commented:
Hey @spen - I've added the requested changes.

I've also added a Test plan. The only bit I haven't covered is the site importer in those test steps, since I'm not sure how to get it working. Any advice?

dmsnell commented on client/my-sites/importer/importing-pane.jsx:
```diff
@@ -254,11 +261,21 @@ class ImportingPane extends React.PureComponent {
+	return {
+		...boundActionCreators,
+		mapAuthorFor: importerId => ( source, target ) =>
+			setTimeout( () => {
```
why are we deferring at all?

bisko replied:
That's a leftover from the Flux flow :/ It shouldn't be needed now?

spen replied:
I'm not sure, ideally we wouldn't be. I remember there being some conversation around this though, @bisko did you know the reasons for this pattern?

dmsnell replied:
If you look at some of the comments around those original `defer()` statements they are meant to indicate that the `defer()` should be removed in a Redux refactor

dmsnell commented on client/state/imports/utils.js:
```diff
@@ -5,7 +5,6 @@
  */
 
 import { find } from 'lodash';
-import { fromJS } from 'immutable';
```
historical note: this was written during a time of early exploration on how to implement immutable behaviors in Calypso. it's good to remove it.

dmsnell commented on client/my-sites/importer/importing-pane.jsx:
```diff
@@ -193,6 +196,10 @@ class ImportingPane extends React.PureComponent {
 
+	startImporting = () => {
+		this.props.startImporting( this.props.importerStatus );
+	};
```
fun fact! we can skip the clutter with a single expression

```js
startImporting = () => this.props.startImporting( this.props.importerStatus );
```

talldan replied:
👍 will shorten.

dmsnell commented on client/my-sites/importer/importing-pane.jsx:
```diff
@@ -254,11 +261,21 @@ class ImportingPane extends React.PureComponent {
+				dispatch( mapAuthor( importerId, source, target ) );
+			} ),
+	};
+};
```
A coupled notes on this here…

I'm curious about our use of `bindActionCreators` because I'm not sure it's adding any value or making the code any clearer. Is there a reason not to return the simpler `startImporting: () => dispatch( startImporting )`?

Second, what's the need for the `defer()`/`setTimeout()` here? I don't think it should be necessary. Without it our function can turn into the shorthand syntax as well…

```js
const mapDispatchToProps = dispatch => ( {
	startImporting: () => dispatch( startImporting() ),
	mapAuthorFor: importerId => ( source, target ) => dispatch( mapAuthor( importerId, source, target ) ),
} )
```

talldan replied:
For the `defer` - this might have been related to flux, and some weirdness around dispatching multiple actions. This comment would seem to indicate that's the case:
https://github.com/Automattic/wp-calypso/pull/23930#discussion_r179778504

I can look to simplify the `mapDispatchToProps`. One way could be that the `mapAuthorFor` higher order function becomes a method of the class:
```
mapAuthorFor: importerId => ( source, target ) => this.props.mapAuthor( importerId, source, target );
```
then I think I can use the shorthand for mapDispatchToProps and inline it:
```
export default connect( null, {
  startImporting,
  mapAuthor,
} )( localize( ImportingPane ) );
```

dmsnell commented on client/my-sites/importer/site-importer/site-importer-input-pane.jsx:
```diff
@@ -233,11 +241,23 @@ class SiteImporterInputPane extends React.Component {
+			defer( () => {
+				dispatch( mapAuthor( importerId, source, target ) );
+			} ),
+	};
```
same questions as above

dmsnell commented on client/state/imports/reducer.js:
```diff
@@ -0,0 +1,290 @@
+ * @param {Object} apiUpdate The object to merge into the api object
+ * @returns {Object} The updated state with api changes
+ */
+const mergeApiUpdate = curry( ( apiUpdate, state ) => ( {
```
I think we'd like to avoid the currying syntax here. This is some pretty old code and I think we'll want to consider updating it instead of just moving it over into Redux, and hopefully that's not a massive chore switching paradigms.

That is, it would be nice to see the functions rewritten without `curry()`, preferring `apiUpdate => state => …` syntax or even better, uncurried with `( state, apiUpdate )` args, for example

talldan replied:
👍 will make those changes

dmsnell commented on client/state/imports/reducer.js:
```diff
@@ -0,0 +1,290 @@
+			...state.importerLocks,
+			[ importerId ]: false,
+		},
+	} ),
```
With `immutable-js`'s built-in helper functions having a single large state object is a bit easier, but here we end up with lots of duplication with the spread operator and one big reducer. We prefer building our normal reducers with composition these days with `combineReducers()` and allow for each sub-reducer to handle its own input. it would be nice to explore this idea now while rewriting all the state code

talldan replied:
Yeah, I agree the state/reducers could do with a bit of a revamp now that they're in redux.

I thought it might be biting off a bit much to tackle all that in the one PR (I'm a trialist, so fairly new to the codebase).

dmsnell replied:
oh that's fine - actually I was hoping to suggest that we break this PR down - it's much much bigger than we usually like to merge, and I'm always advocating for finding ways to refactor things gradually and slowly instead of all-at-once.

that is, maybe we can find small ways to start updating this code in pieces, merge them one at a time, give each change a due chance for review and testing, then move to the next one…

talldan replied:
It'd be quite hard to migrate the state piece by piece, but the PR could be divided up into one PR for actions, one for reducers, and a third and final PR to hook it all up (and remove the old code). That's kind of how I did my commits anyway.

It would make it more reviewable with the only downside being that the first two PRs would add unused code that couldn't receive any manual tests (but at least it'll be code with unit tests).

dmsnell replied:
did you consider breaking apart the state structure and porting over parts of the state at a time?

for example, one PR to create the basic reducer and include in it one data item and the action types relevant to it. that is, create some _temporary_ duplication in the code in order to gradually shift the data structure

```js
this.props.oldFluxyUpdate();
this.props.update() // new redux action dispatcher
```

talldan replied:
If I'm understanding the proposal, I think it could be complicated by the way the imports state is structured. Many of the action creators trigger multiple actions and the reducers read from and update multiple parts of the importer state, they're not isolated in a way that would make it easy to have some implemented in flux and some in redux.

I think the options to proceed along those lines would be either:
a) refactor the state so that it can be migrated piece by piece (and abandoning this PR)
b) to have some way to make flux read from particular parts of redux's state tree (sounds scary, but maybe you've used similar approaches in the past?) and breaking this PR up.

Then again, I might be completely misunderstanding.

dmsnell replied:
1500 lines isn't unheard of, but it's not really about the size so much as the surface area for bugs and the ability to review and find them.

option (a) sounds legitimate; this PR could be a good guide for splitting up the work. ideally it would be great if we could propose a PR that makes one single change and needs one specific thing to test, verses one that changes an entire subsystem and whose testing instructions are "make sure imports still work"

in this case maybe the risk is lower because of the exist code? maybe not? I was hoping that we could pull out something, for example the author mapping, into a separate reducer, etc… and move gradually.

anyway, thanks for evaluating this. I'd still propose the challenge and suggest you explore it a little bit on how possible it is to break apart, but if you end up thinking it will take three times as long (for example) then let pragmatism be your guide. just remember that usually we underestimate how long single large atomic changes can take :wink:

samouri replied:
> to have some way to make flux read from particular parts of redux's state tree

we have a module called the [redux-bridge](https://github.com/Automattic/wp-calypso/blob/master/client/lib/redux-bridge/index.js)

talldan replied:
Thanks for the pointer - I'm just looking into this to see if there's a way I can break this up into incremental PRs.

jsnajdr replied:
@talldan I do similar migrations as a series of smaller PRs that go like this:
1. Add the Redux store with actions, reducers and selectors. It's not used anywhere yet. It's tested only by unit tests. It's usually a large diff, but it's not dangerous if some bug is not catched in this phase, as the code is not yet used.
2. Use the new Redux store instead of Flux in the app code. This phase needs to be reviewed and tested most carefully, and it helps that the diff is usually rather small.
3. Remove the old Flux code that's no longer used. Large diff, but trivial to review.

The most important benefit is that the reviewer doesn't need to scroll through large diffs like the current PR. The risky part is concentrated in a small diff in part 2.

dmsnell replied:
I'd like to add that it's nice to break step (1) into small pieces as well, and if possible, run several flows through 1/2/3 :smile:

dmsnell commented on client/state/imports/utils.js:
```diff
@@ -57,11 +56,7 @@ function replaceUserInfoWithIds( customData ) {
-							mappedTo: author.mappedTo.ID,
-						} )
-					: author
+				author.mappedTo ? Object.assign( {}, author, { mappedTo: author.mappedTo.ID } ) : author
```
feel free to use shorter notation - no need to be verbose here

```js
author.mappedTo ? { ...author, mappedTo: author.mappedTo.ID } : author
```

talldan replied:
👍 will make that change

dmsnell commented:
@talldan on note on my review comments here - I may sound like I have strong opinions (and I do) but also please take the into consideration with everything else. I will try to be clear if there is something I really want to make sure we do but most of my comments are meant to guide or question and let you decide.

jsnajdr commented on client/state/imports/actions.js:
```diff
@@ -55,38 +54,39 @@ const expiryOrder = ( siteId, importerId ) =>
+const apiStart = () => ( { type: IMPORTS_FETCH } );
+
+const apiSuccess = dispatch => data => {
+	dispatch( { type: IMPORTS_FETCH_COMPLETED } );
```
`apiSuccess`, `apiFailure` and `receiveImporterStatus` should be classic action creators: function that returns an action object (with a `type` field etc.).

jsnajdr commented on client/state/imports/actions.js:
```diff
@@ -98,32 +98,33 @@ export function cancelImport( siteId, importerId ) {
-		.catch( apiFailure );
-}
+		.then( receiveImporterStatus( dispatch ) )
+		.catch( apiFailure( dispatch ) );
```
With classic action creators, this code can become:
```js
wpcom.updateImporter( ... ).then( data => {
  dispatch( apiSuccess() );
  dispatch( receiveImporterStatus( fromApi( data ) ) );
} ).catch( error => {
  dispatch( apiFailure() );
} );
```

As a next step, can we merge the `IMPORTS_FETCH_COMPLETED` and `IMPORTS_IMPORT_RECEIVE` actions into one? They are always dispatched together.

dmsnell replied:
…and we can skip all this by using `trackRequests` request-tracking

jsnajdr commented on client/state/imports/actions.js:
```diff
@@ -55,38 +54,39 @@ const expiryOrder = ( siteId, importerId ) =>
 const lockImport = partial( setImportLock, true );
 const unlockImport = partial( setImportLock, false );
 
 const asArray = a => [].concat( a );
```
There's `lodash/castArray` that does exactly the same thing.

jsnajdr commented on client/state/imports/actions.js:
```diff
@@ -55,38 +54,39 @@ const expiryOrder = ( siteId, importerId ) =>
+	return { type, importerId };
 };
 const lockImport = partial( setImportLock, true );
 const unlockImport = partial( setImportLock, false );
```
If these two functions were just
```js
const lockImport = importerId => ( { type: IMPORTS_IMPORT_LOCK, importerId } );
const unlockImport = importerId => ( { type: IMPORTS_IMPORT_UNLOCK, importerId } );
```
they would be admittedly more boring, but maybe more readable and shorter.

jsnajdr commented on client/my-sites/site-settings/section-import.jsx:
```diff
@@ -80,11 +73,7 @@ class SiteSettingsImport extends Component {
-
-	updateState = () => {
-		this.setState( getImporterState() );
+		this.props.fetchState( this.props.site.ID );
```
Can we create a `<QueryImporterState siteId />` component for this? That's how we trigger data fetching in components.

Also, guard for `this.props.site` being `null`. That's a pre-existing bug that can crash the `/settings/import/:site_id` page when the sites list is not yet loaded.

jsnajdr commented on client/state/imports/actions.js:
```diff
@@ -220,10 +228,4 @@ export const startUpload = ( importerStatus, file ) => dispatch => {
 		.then( fromApi )
```
The `onabort` handler should be `() => dispatch( cancelImport( siteId, importerId ) )`

jsnajdr commented on client/state/imports/actions.js:
```diff
@@ -220,10 +228,4 @@ export const startUpload = ( importerStatus, file ) => dispatch => {
 		.then( fromApi )
 		.then( flowRight( dispatch, finishUpload( importerId ) ) )
 		.catch( flowRight( dispatch, failUpload( importerId ) ) );
```
Avoiding `flowRight` and similar functions generally leads to more readable code:
```js
then( importerStatus => dispatch( finishUpload( importerId, importerStatus ) ) )
```

jsnajdr commented on client/state/imports/reducer.js:
```diff
@@ -0,0 +1,290 @@
+				retryCount: state.api.retryCount + 1,
+			},
+			state
+		),
```
Can `state => mergeApiUpdate( _, state )` be replaced with `mergeApiUpdate( _ )`?

jsnajdr commented on client/state/imports/reducer.js:
```diff
@@ -0,0 +1,290 @@
+ * @param {Object} state The state of the redux store
+ * @returns {Object} state The updated state with count incremented by 1
+ */
+const incrementCount = state => ( { ...state, count: state.count + 1 } );
```
To make reducers simpler, we often slice them into individual pieces and combine them with `combineReducers`. For example, the `count` reducer could look like this:
```js
const count = ( state = 0, action ) => {
  switch ( action.type ) {
    case IMPORTS_STORE_RESET:
      return 0;
    case IMPORTS_IMPORT_START:
      return state + 1;
  }
  return state;
}

export default combineReducers( count, ... );
```
Very clean, isn't? And it can also avoid many nested updates.

talldan commented:
@dmsnell @jsnajdr Based on the feedback, today I've had a go at incrementally migrating part of the store state (`importerLocks`). There's a PR here:
https://github.com/Automattic/wp-calypso/pull/24679

I used the redux-bridge (thanks @samouri)—it seems to work pretty well, and has resulted in what is hopefully a much more palatable PR.

alisterscott commented:
@talldan are we able to close this one?

talldan commented:
@alisterscott I've closed this one now, sorry for the lack of response.
